### ISSUE TITLE
Add traceCapable indication

### ISSUE THREAD

**Issue by nickelpro** [verilator/verilator#5053]
We recently bumped our Verilator build and ran head first into #4462, which adds a tracing method to models even when no tracing is enabled in the build.

In a classic case of Hyrum's Law, [this broke our build](https://xkcd.com/1172/).

We setup our test framework to automagically generate coverage reports, traces, and other test artifacts depending on the capabilites of the given Verilated model, that way the same test can be used to generate various data depending on how the model is generated.

Basically we have something like this:

```cpp
template <typename T>
concept traceable = requires(T a, VerilatedFstC fst) { a.trace(&fst, 1); };

template <typename T>
concept not_traceable = evalable<T> && !traceable<T>;
```

And various functions with different overloads depending on if model is traceable or not (among other things).

I don't have any objection to #4462, but for my own sanity I would appreciate a flag to hard remove the tracing method/restore the old behavior.

**Comment by wsnyder** [comment]
Note the master version now has trace() as part of the context, the old function just calls into that.  

Would you like to propose a pull to do what you need?

**Comment by nickelpro** [comment]
I can pull it but I'm not a perl programmer and the testing system seems involved, so I wouldn't trust myself to write useful tests.

The title of this issue is bad, what I want is _any_ compile-time mechanism to determine if a given CModel was verilated with tracing enabled. The native C++ mechanism for this would probably be a trait class, so something like:

```cpp
// Somewhere in the vicinity of verilated.h
template<typename T>
struct VerilatedTraceable : std::false_type {};

#if VERILATOR_USE_CONCEPTS
template<typename T>
concept VerilatedIsTraceable = VerilatedTraceable<T>::value;
#endif

// V3EmitCModel.cpp
if(v3Global.opt.trace()) puts("template<> struct VerilatedTraceable<" + topClassName() + "> : std::true_type {};");
```
The goal being to enable test code like so:
```cpp
void my_custom_eval_function(VerilatedIsTraceable auto& DUT) {
  DUT.eval();
  // ...
  FST->dump(timestamp);
}

void my_custom_eval_function(auto& DUT)  {
  DUT.eval();
  // Don't try to generate traces or dump()
}
```

**Comment by wsnyder** [comment]
What prevents using `#ifdef VM_TRACE`?

**Comment by nickelpro** [comment]
Too broad a brush, `VM_TRACE` tells me at least one model in the translation unit was verilated with `--trace`, it does not tell me which ones.

We might put a hundred such models inside a single Catch2 binary, and frequently a dev will only turn on tracing for one or two of them that they're currently debugging.

**Comment by wsnyder** [comment]
Would a 'constexpr bool traceEnabled = true/false" within the model's class be sufficient?

**Comment by wsnyder** [comment]
Please give it a try.

**Comment by nickelpro** [comment]
Perfect, much appreciated
